libs/core/src/bloom_filter.cpp

```diff
@@ -0,0 +1,375 @@
+{
+  return HashSource::HashSourceIterator{this, 0u};
+}
+
+HashSource::HashSourceIterator HashSource::cend() const
+{
+  return HashSource::HashSourceIterator{this, data_.size()};
+}
+
+std::size_t HashSource::getHash(std::size_t index) const
+{
+  return data_[index];
+}
+
+HashSource::HashSourceIterator::~HashSourceIterator() = default;
+
+HashSource::HashSourceIterator::HashSourceIterator(HashSourceIterator const &)     = default;
+HashSource::HashSourceIterator::HashSourceIterator(HashSourceIterator &&) noexcept = default;
+
+HashSource::HashSourceIterator &HashSource::HashSourceIterator::operator=(
+    HashSource::HashSourceIterator const &) = default;
+HashSource::HashSourceIterator &HashSource::HashSourceIterator::operator=(
+    HashSource::HashSourceIterator &&) noexcept = default;
+
+bool HashSource::HashSourceIterator::operator!=(HashSource::HashSourceIterator const &other) const
+{
+  return !(operator==(other));
+}
+
+bool HashSource::HashSourceIterator::operator==(HashSource::HashSourceIterator const &other) const
+{
+  return source_ == other.source_ && hash_index_ == other.hash_index_;
+}
+
+HashSource::HashSourceIterator const HashSource::HashSourceIterator::operator++(int)
+{
+  HashSource::HashSourceIterator temp{source_, hash_index_};
+
+  operator++();
+
+  return temp;
+}
+
+HashSource::HashSourceIterator &HashSource::HashSourceIterator::operator++()
+{
+  ++hash_index_;
+
+  return *this;
+}
+
+std::size_t HashSource::HashSourceIterator::operator*() const
+{
+  return source_->getHash(hash_index_);
+}
+
+HashSource::HashSourceIterator::HashSourceIterator(HashSource const *source, std::size_t index)
+  : hash_index_{index}
+  , source_{source}
+{}
+
+namespace {
+
+class OpenSslHasher
+{
+public:
+  enum class Type
+  {
+    MD5,
+    SHA_1,
+    SHA_2_512,
+    SHA_3_512
+  };
+
+  explicit OpenSslHasher(Type type)
+    : openssl_type_{to_openssl_type(type)}
+    , ctx_{EVP_MD_CTX_create()}
+  {
+    EVP_MD_CTX_init(ctx_);
+  }
+
+  ~OpenSslHasher()
+  {
+    EVP_MD_CTX_destroy(ctx_);
+  }
+
+  void reset() const
+  {
+    EVP_DigestInit_ex(ctx_, openssl_type_, nullptr);
+  }
+
+  std::vector<std::size_t> operator()(HashSourceFactory::Bytes const &input) const
+  {
+    const auto size_in_bytes = static_cast<std::size_t>(EVP_MD_CTX_size(ctx_));
+
+    std::vector<std::size_t> output((size_in_bytes + sizeof(std::size_t) - 1) /
+                                    sizeof(std::size_t));
+    EVP_DigestUpdate(ctx_, input.pointer(), input.size());
+
+    EVP_DigestFinal_ex(ctx_, reinterpret_cast<uint8_t *>(output.data()), nullptr);
+
+    return output;
+  }
+
+private:
+  EVP_MD const *to_openssl_type(Type const type) const
+  {
+    EVP_MD const *openssl_type = nullptr;
+    switch (type)
+    {
+    case Type::MD5:
+      openssl_type = EVP_sha512();
+      break;
+    case Type::SHA_1:
+      openssl_type = EVP_sha1();
+      break;
+    case Type::SHA_2_512:
+      openssl_type = EVP_sha512();
```

Comment:
Can we not add these hashes to our existing crypto library?

Reply:
I was thinking of doing it as a follow up, once we decide which ones we actually need.